### PR TITLE
Commands/network IT2

### PR DESCRIPTION
**Pull request by BurtBiel**
```
Command
    az network vpn-gateway create

Arguments
    --name -n           [Required]: Gateway name.
    --public-ip-address [Required]: Name or ID of the resource.  Uses existing by default.
    --resource-group -g [Required]: Name of resource group.
    --subnet-id         [Required]: Subnet ID for the VPN gateway. Subnet must be named
                                    GatewaySubnet.
    --enable-bgp                  : Enable BGP (Border Gateway Protocol).
    --gateway-type                : Allowed values: ExpressRoute, Vpn.  Default: Vpn.
    --location -l                 : Location.
    --sku                         : Allowed values: Basic, HighPerformance, Standard.  Default:
                                    Basic.
    --tags                        : Multiple semicolon separated tags in 'key[=value]' format.  Use
                                    "" for none.
    --vpn-type                    : Allowed values: PolicyBased, RouteBased.  Default: RouteBased.
```

```
Command
    az network vpn-connection create

Arguments
    --connection-type   [Required]: Allowed values: ExpressRoute, IPSec, Vnet2Vnet.
    --name -n           [Required]: Connection name.
    --resource-group -g [Required]: Name of resource group.
    --vnet-gateway1-id  [Required]: Connect from this gateway to another gateway or express route
                                    circuit.
    --express-route-circuit2      : Connect to this express route circuit from vnet gateway 1 using
                                    connection type ExpressRoute.
    --local-gateway2-id           : Connect to this local gateway from vnet gateway 1 using
                                    connection type IPSec.
    --location -l                 : Location.
    --routing-weight              : Connection routing weight.  Default: 10.
    --shared-key                  : IPSec shared key.
    --vnet-gateway2-id            : Connect to this vnet gateway from vnet gateway 1 using
                                    connection type Vnet2Vnet.
```

```
Command
    az network route-table route create

Arguments
    --address-prefix    [Required]: Gets or sets the destination CIDR to which the route applies.
    --name -n           [Required]: Route name.
    --next-hop-type     [Required]: Gets or sets the type of Azure hop the packet should be sent to.
                                    Allowed values: Internet, None, VirtualAppliance,
                                    VirtualNetworkGateway, VnetLocal.
    --resource-group -g [Required]: Name of resource group.
    --route-table-name  [Required]: Route table name.
    --next-hop-ip-address         : Gets or sets the IP address packets should be forwarded to. Next
                                    hop values are only allowed in routes where the next hop type is
                                    VirtualAppliance.
```

```
Command
    az network vpn-gateway root-cert create

Arguments
    --gateway-name      [Required]: Virtual network gateway name.
    --name -n           [Required]: Root certificate name.
    --public-cert-data  [Required]: Base64 contents of the root certificate file or file path.
    --resource-group -g [Required]: Name of resource group.
```

... rest follow similar pattern

### CODE REVIEW

File: src/azure/cli/commands/arm.py [hunk 1]
```diff
@@ -172,32 +172,39 @@ def handler(args):
                       if key in get_arguments}
         instance = getter(client, **getterargs) if client else getter(**getterargs)
 
+        for k in args.copy().keys():
+            if k in get_arguments or k in set_arguments \
+                or k in ('properties_to_add', 'properties_to_remove', 'properties_to_set'):
+                args.pop(k)
+        for key, val in args.items():
+            ordered_arguments.append((key, val))
```
Comment (tjprescott) [inline src/azure/cli/commands/arm.py]: Out of curiosity, why is this here? What was the problem the way it was before?

File: src/azure/cli/commands/arm.py [hunk 2]
```diff
@@ -228,8 +231,7 @@ def one_required(namespace):
                      metavar='LIST KEY=VALUE')
     cmd.add_argument('properties_to_remove', '--remove', nargs='+', action=OrderedArgsAction,
                      default=[], help='Remove a property or an element from a list.  Example: '
-                     '--remove property.list <index>', metavar='LIST INDEX',
-                     validator=one_required)
+                     '--remove property.list <index>', metavar='LIST INDEX')
```
Comment (tjprescott) [inline src/azure/cli/commands/arm.py]: >LIST INDEX [](start = 64, length = 10)

I believe metavar is supposed to be one word or use _ instead of space to avoid confusing the user who might interpret it as needing two values.

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/custom.py [hunk 3]
```diff
@@ -718,4 +719,92 @@ def update_subnet(resource_group_name, virtual_network_name, subnet_name,
...
+                       ' certificates.  Please use "update" with --address-prefixes first.')
+
+    if config.vpn_client_revoked_certificates is None:
+        config.vpn_client_revoked_certificates = []
+    if config.vpn_client_root_certificates is None:
+        config.vpn_client_root_certificates = []
+
+    return config, gateway, ncf
+
+def handle_address_prefixes(instance, _, arg_value):
```
Comment (tjprescott) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/custom.py]: > handle_address_prefixes [](start = 3, length = 24)

Nit: should this be prepended with _? It looks like an internal helper method.

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/_params.py [hunk 4]
```diff
@@ -194,15 +200,20 @@ def completer(prefix, action, parsed_args, **kwargs): # pylint: disable=unused-a
...
-register_cli_argument('network express-route circuit-peering', 'peering_name', name_arg_type)
-register_cli_argument('network express-route circuit', 'circuit_name', name_arg_type, completer=get_resource_name_completion_list('Microsoft.Network/expressRouteCircuits'))
+register_cli_argument('network express-route circuit-auth', 'authorization_name', name_arg_type, id_part='child_name', help='Authorization name')
+register_cli_argument('network express-route circuit-auth', 'circuit_name', completer=get_resource_name_completion_list('Microsoft.Network/expressRouteCircuits'), id_part='name', help='Circuit name.')
+register_cli_argument('network express-route circuit-auth', 'authorization_key', help='Authorization key.')
 
 # Express Route Circuit Peering
 register_cli_argument('network express-route circuit-peering', 'peering_name', name_arg_type, id_part='child_name')
+register_cli_argument('network express-route circuit-peering', 'peering_type', choices=get_enum_choices(peeringType))
+register_cli_argument('network express-route circuit-peering', 'sku_family', choices=get_enum_choices(sku_family))
```
Comment (tjprescott) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/_params.py]: >, choices=get_enum_choices(sku_family) [](start = 75, length = 38)

Are these enums modeled as strings in Swagger? If so, per Johan's guidance, we should use an enum completer instead of argparse choices, and put example values in the help text.

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/_params.py [hunk 5]
```diff
@@ -339,9 +351,15 @@ def completer(prefix, action, parsed_args, **kwargs): # pylint: disable=unused-a
 register_cli_argument('network vpn-gateway create', 'sku', choices=get_enum_choices(sku))
 register_cli_argument('network vpn-gateway create', 'vpn_type', choices=get_enum_choices(vpnType))
 register_folded_cli_argument('network vpn-gateway create', 'public_ip_address', 'Microsoft.Network/publicIPAddresses', default_value_flag='existingId', allow_none=False, required=True)
+register_cli_argument('network vpn-gateway', 'cert_name', help='Root certificate name', options_list=('--name', '-n'))
+register_cli_argument('network vpn-gateway', 'gateway_name', help='Virtual network gateway name')
+register_cli_argument('network vpn-gateway root-cert create', 'public_cert_data', help='Base64 contents of the root certificate file or file path.', validator=load_cert_file)
```
Comment (tjprescott) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/_params.py]: >Base64 contents of the root certificate file or file path. [](start = 88, length = 58)

From the validator, it looks like you are loading a cert file. Thus this should be aliased to something like (--cert-path) and the help text changed as appropriate. This is what I did for app-gateway so that it appears correct for CLI users as well as anyone that might be using the smart create (where they really would pass in the data because that's what the template requires)

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/_validators.py [hunk 6]
```diff
@@ -263,6 +267,9 @@ def process_ag_url_path_map_create_namespace(namespace): # pylint: disable=unuse
         namespace.default_http_settings = _generate_ag_subproperty_id(
             namespace, 'backendHttpSettingsCollection', namespace.default_http_settings)
 
+        # extract the certificate data from the provided file
+        namespace.cert_data = read_base_64_file(namespace.cert_data)
+
```
Comment (tjprescott) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/_validators.py]: Why is this here? App-gateway url-path-map create does not have a cert_data field. This should crash.

**Comment by tjprescott** [comment]
:clock1:

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/_validators.py [hunk 7]
```diff
@@ -308,6 +315,10 @@ def process_public_ip_create_namespace(namespace):
     if namespace.dns_name:
         namespace.public_ip_address_type = 'dns'
 
+def load_cert_file(namespace):
+    if os.path.isfile(namespace.public_cert_data):
```
Comment (tjprescott) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/_validators.py]: > if os.path.isfile(namespace.public_cert_data): [](start = 3, length = 47)

Why this check? If the specified path isn't a file, why fail silently? Why not let it throw?

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/mgmt_express_route_circuit/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        "description": "Name of the ExpressRoute Service Provider. It must exactly match one of the Service Providers from List ExpressRoute Service Providers API call."
+      }
+    },
+    "sku_family": {
+      "type": "string",
+      "defaultValue": "MeteredData",
+      "allowedValues": [
+        "MeteredData",
+        "UnlimitedData"
+      ],
```
Comment (tjprescott) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/mgmt_express_route_circuit/azuredeploy.json]: Right, so these enums are set to "ModelAsString=True" in the Swagger. Thus: 1) they should not have allowedValues in the ARM template. 2) They should use completers instead of choices in the parameter aliasing. 3) The help text should list some or all of the "allowedValues".  I've already got an issue opened against the Swagger... feel free to vote it up!!

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/mgmt_express_route_peering/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+      }
+    },
+    "peeringType": {
+      "type": "string",
+      "defaultValue": "AzurePrivatePeering",
+      "allowedValues": [
+        "AzurePrivatePeering",
+        "AzurePublicPeering",
+        "MicrosoftPeering"
+      ],
```
Comment (tjprescott) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/mgmt_express_route_peering/azuredeploy.json]: Same enum comment here.

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/mgmt_vpn_connection/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+        "description": "Connection name."
+      }
+    },
+    "connectionType": {
+      "type": "string",
+      "allowedValues": [
+        "IPSec",
+        "Vnet2Vnet",
+        "ExpressRoute"
+      ],
```
Comment (tjprescott) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/mgmt_vpn_connection/azuredeploy.json]: More enums that need to be converted.

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/tests/test_network_commands.py [hunk 11]
```diff
@@ -863,50 +859,51 @@ def body(self):
...
+
+        subnet1_id = '/subscriptions/{subscription_id}/resourceGroups' \
+            '/{rg}/providers/Microsoft.Network/virtualNetworks/{vnet1}/subnets/GatewaySubnet' \
+            .format(subscription_id=subscription_id, rg=rg, vnet1=vnet1_name)
+        subnet2_id = '/subscriptions/{subscription_id}/resourceGroups' \
+            '/{rg}/providers/Microsoft.Network/virtualNetworks/{vnet2}/subnets/GatewaySubnet' \
+            .format(subscription_id=subscription_id, rg=rg, vnet2=vnet2_name)
+
+        self.cmd('network vpn-gateway create -g {0} -n {1} --subnet-id {2} --public-ip-address {3}'
+                 .format(rg, gateway1_name, subnet1_id, ip1_name))
```
Comment (tjprescott) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/tests/test_network_commands.py]: >) [](start = 65, length = 1)

Why no checks (either JMESPath checks or subsequent commands) to verify the expected result?

**Comment by tjprescott** [comment]
vpn-gateway create > public-ip-address: What does uses existing by default mean? Is there a way to create a new one? If not, why not just say "name or ID of existing public ip" (why does it just generically say resource?)

**Comment by tjprescott** [comment]
There's the 'use "" for none' on tags. I think you said in the last PR that you might address it in a future PR since it affects more that just this, but I forget. :o

**Comment by tjprescott** [comment]
Since Yugang is OOF, I'll take up the mantel of comparing this to Xplat. :) So, looking at the help contents for the commands, some comments:

vpn-connection create: Xplat uses lnet-gateway instead of local-gateway. Any reason we shouldn't follow suit?

**Comment by tjprescott** [comment]
vpn-connection > shared-key: Is this sensitive info that you wouldn't want being recorded in the command history/logs/etc?  If so, we might need to accept a path to a file containing the shared-key instead of the raw key itself.

**Comment by tjprescott** [comment]
vpn-connection > connection-type: can we determine the correct connection type by looking at the supplied parameters?

**Comment by BurtBiel** [comment]
this text is generic as a part of param folding.  I updated the text to be a little clearer

---
In reply to: [239469482](https://github.com/Azure/azure-cli/pull/687#issuecomment-239469482) [](ancestors = 239469482)

**Comment by BurtBiel** [comment]
I updated the text to be a little clearer

---
In reply to: [239469535](https://github.com/Azure/azure-cli/pull/687#issuecomment-239469535) [](ancestors = 239469535)

**Comment by BurtBiel** [comment]
local gateway is the terminology used by the portal and documentation, so I vote for clarity over brevity in this case

---
In reply to: [239469581](https://github.com/Azure/azure-cli/pull/687#issuecomment-239469581) [](ancestors = 239469581)

**Comment by BurtBiel** [comment]
added file loading to this param

---
In reply to: [239469625](https://github.com/Azure/azure-cli/pull/687#issuecomment-239469625) [](ancestors = 239469625)

File: src/azure/cli/commands/arm.py (hunk 1 above)
Comment (BurtBiel) [inline src/azure/cli/commands/arm.py]: ordered args previously only took the set, add and remove params.  Now we add additional params so they will be processed by the custom handlers

---
In reply to: [74598948](https://github.com/Azure/azure-cli/pull/687#discussion_r74598948) [](ancestors = 74598948)

File: src/azure/cli/commands/arm.py (hunk 2 above)
Comment (BurtBiel) [inline src/azure/cli/commands/arm.py]: We've used multi-word metavars for some other params, Johan was onboard with those changes

---
In reply to: [74599192](https://github.com/Azure/azure-cli/pull/687#discussion_r74599192) [](ancestors = 74599192)

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/_params.py (hunk 4 above)
Comment (BurtBiel) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/_params.py]: no, if they were model as string there wouldn't be a strongly typed enum, so in this case choices make sense

---
In reply to: [74600413](https://github.com/Azure/azure-cli/pull/687#discussion_r74600413) [](ancestors = 74600413)

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/_params.py (hunk 5 above)
Comment (BurtBiel) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/_params.py]: This field can take either a file or a path, so cert-path would rule out the raw option

---
In reply to: [74600850](https://github.com/Azure/azure-cli/pull/687#discussion_r74600850) [](ancestors = 74600850)

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/_validators.py (hunk 7 above)
Comment (BurtBiel) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/_validators.py]: this supports having raw data passed in on the command line, so we leave the data alone if it is not a file 

---
In reply to: [74602269](https://github.com/Azure/azure-cli/pull/687#discussion_r74602269) [](ancestors = 74602269)

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/_validators.py (hunk 6 above)
Comment (BurtBiel) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/_validators.py]: Bad auto-merge I believe.  Removed 

---
In reply to: [74601604](https://github.com/Azure/azure-cli/pull/687#discussion_r74601604) [](ancestors = 74601604)

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/tests/test_network_commands.py (hunk 11 above)
Comment (BurtBiel) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/tests/test_network_commands.py]: the command will fail if anything goes wrong here, if it didn't, the following command will fail when it tries to connect the two gateways.  If the second command passes but didn't create properly, then we have a bigger problem on our hands :)

---
In reply to: [74603966](https://github.com/Azure/azure-cli/pull/687#discussion_r74603966) [](ancestors = 74603966)

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/mgmt_vpn_connection/azuredeploy.json (hunk 10 above)
Comment (BurtBiel) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/mgmt_vpn_connection/azuredeploy.json]: this is a template switch variable, so it can't have arbitrary values

---
In reply to: [74603489](https://github.com/Azure/azure-cli/pull/687#discussion_r74603489) [](ancestors = 74603489)

**Comment by BurtBiel** [comment]
@tjprescott comments addressed
